### PR TITLE
Add a grpclb-in-DNS interop test suite

### PR DESCRIPTION
**Pull request by apolcyn**
This PR adds a test suite that can bring up fake and local grpclb servers, backends, fallback servers, and a controlled DNS server. It currently runs grpc interop clients against the servers. The test driver takes a list of scenarios that tell it the number of "backends", "fallback" servers, and "grpclb" servers to bring up, what credentials to listen on, and certain behaviors to provide. For each scenario, it brings up all of the fake servers in docker containers, wires up the DNS records of the DNS server to point to the IP addresses of each docker container, and then runs the clients. Note also this currently only tests the native C++, Java, and Go stacks, but it could be extended.

General things that this PR does:

1) Create `tools/run_tests/run_grpclb_interop_tests.py`, which is the main test driver that basically follows the description above.

2) Add `tools/run_tests/lb_interop_tests/gen_build.yaml`, which generates a list of "test scenario" JSON configs into `tools/run_tests/generated/lb_interop_test_scenarios.json`

3) Add a new docker file under `tools/run_tests/dockerfile/lb_interop_fake_servers/` which has the dependencies needed to build the fake backend, fallback, and balancer server, and to run the local python DNS server (the balancer is from https://github.com/grpc/grpc-go/pull/2338)

4) Make a couple of tweaks to existing interop client builds to add logging and compile faster

See: https://source.cloud.google.com/results/invocations/f93c0e2c-975d-48af-8464-1d86488ad505/targets for a passing run

These new tests exposed a few bugs in a few scenarios:
* https://github.com/grpc/grpc-go/issues/2347
* https://github.com/grpc/grpc-java/issues/4887
* https://github.com/grpc/grpc-java/issues/4912

.... and so those scenarios are currently being skipped.

Remaining TODO is to use a non-experimental kokoro config file, otherwise I think this is ready for review

-----------------------------

cc @jtattermusch can you please review for general sanity on adding this new suite?

cc @zhangkun83  can you please review, in particular, for sanity of the devised scenarios here? Scenarios are in `tools/run_tests/lb_interop_tests/gen_build_yaml.py` . . . happy to also talk offline here.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,802      Total (=)      1,964,802

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,466      Total (<)     10,887,473

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,802      Total (=)      1,964,802

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,471      Total (>)     10,887,469

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,792      Total (=)      1,964,792

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,069      Total (>)     10,887,065

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,792      Total (=)      1,964,792

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,063      Total (<)     10,887,065

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,792      Total (=)      1,964,792

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,064      Total (<)     10,887,068

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,792      Total (=)      1,964,792

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,064      Total (=)     10,887,064

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,792      Total (=)      1,964,792

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,061      Total (<)     10,887,070

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,792      Total (=)      1,964,792

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,069      Total (>)     10,887,063

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,892      Total (=)      1,964,892

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,712      Total (<)     10,887,718

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,916      Total (=)      1,964,916

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,712      Total (=)     10,887,712

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: test/cpp/naming/utils/dns_server.py [hunk 1]
```diff
@@ -136,7 +140,14 @@ def main():
                     help='Port for DNS server to listen on for TCP and UDP.')
   argp.add_argument('-r', '--records_config_path', default=None, type=str,
                     help=('Directory of resolver_test_record_groups.yaml file. '
-                          'Defauls to path needed when the test is invoked as part of run_tests.py.'))
+                          'Defauls to path needed when the test is invoked as part '
```
Comment (jtattermusch) [inline test/cpp/naming/utils/dns_server.py]: nit: s/Defauls/Defaults/

File: tools/dockerfile/interoptest/grpc_interop_cxx/build_interop.sh [hunk 2]
```diff
@@ -31,7 +31,7 @@ cd /var/local/git/grpc
 make install-certs
 
 # build C++ interop client & server
-make interop_client interop_server
+make interop_client interop_server -j8
```
Comment (jtattermusch) [inline tools/dockerfile/interoptest/grpc_interop_cxx/build_interop.sh]: not sure about this:
Currently 8 interop images are going to be built in parallel and 8*8 = 64 (the workers have 16 cores).
https://github.com/grpc/grpc/blob/912b8ab4d49cd6cde76675d37c896e5edcf67487/tools/internal_ci/linux/grpc_interop_tocloud.cfg#L29

Change to `-j2` which gives some boost and is conservative enough?

File: templates/tools/dockerfile/java_build_interop.sh.include [hunk 3]
```diff
@@ -25,3 +25,10 @@ cp -r /var/local/jenkins/service_account $HOME || true
 cd /var/local/git/grpc-java
 
 ./gradlew :grpc-interop-testing:installDist -PskipCodegen=true
+
+mkdir -p /var/local/grpc_java_logging
+echo "handlers = java.util.logging.ConsoleHandler
```
Comment (jtattermusch) [inline templates/tools/dockerfile/java_build_interop.sh.include]: nit: add a comment for readability? (`# enable extra java logging`)

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,965,914      Total (=)      1,965,914

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,891,941      Total (=)     10,891,941

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,965,914      Total (=)      1,965,914

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,891,938      Total (=)     10,891,938

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by apolcyn** [comment]
> I think overall, and extra interop script (run_grpclb_interop_tests.py) is going to be pain to maintain and we should do some simplification for all interop tests at some point, but other than that, I don't have any reservations.

Ack, agree some simplifications to the overall interop runners would help.

FWIW I figured this script was different enough from the other `run_interop_tests.py` script though that it made is simpler/easier to put it in its own standalone thing.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,965,918      Total (=)      1,965,918

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,891,799      Total (<)     10,891,800

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,965,918      Total (=)      1,965,918

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,891,799      Total (>)     10,891,788

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,965,918      Total (=)      1,965,918

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,891,796      Total (<)     10,891,797

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,965,918      Total (=)      1,965,918

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,891,798      Total (>)     10,891,795

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,965,922      Total (=)      1,965,922

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,909,005      Total (>)     10,909,003

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,997,294      Total (=)      1,997,294

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,021,760      Total (>)     11,021,756

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

**Comment by apolcyn** [comment]
python test failures: https://github.com/grpc/grpc/issues/15308

**Comment by apolcyn** [comment]
merging, FTR this test will not be running per-PR, at least for now. It will be running continuously on the master branch but will need to be manually monitored for now (if possibly, this can later be promoted to run per-PR, possibly).

**Pull request merged**